src/include/migraphx/op/pack_int4.hpp

```diff
@@ -62,35 +62,62 @@ struct pack_int4
     {
         check_shapes{inputs, *this}.same_dims().has(1);
         auto in_shape = inputs.front();
-        if(in_shape.type() != migraphx::shape::uint8_type)
+        if(in_shape.type() != migraphx::shape::int8_type and
+           in_shape.type() != migraphx::shape::uint8_type)
         {
-            MIGRAPHX_THROW("PACK_INT4: Only Unsigned Int8 type is supported for packing");
+            MIGRAPHX_THROW("PACK_INT4: Only Int8 or Uint8 is supported for packing");
         }
         auto new_lens = in_shape.lens();
         if(new_lens[axis] % 2 != 0)
         {
             MIGRAPHX_THROW("PACK_INT4: Can not pack axis that has odd lengths");
         }
         new_lens[axis] /= 2;
-        return {migraphx::shape::uint8_type, new_lens};
+        return {in_shape.type(), new_lens};
     }
 
     argument compute(const shape& output_shape, std::vector<argument> args) const
     {
+        auto input    = args.front();
+        auto in_shape = input.get_shape();
+
         argument result{output_shape};
-        auto in_shape = args.front().get_shape();
-        auto input    = args.at(0).get<uint8_t>();
-        auto output   = result.get<uint8_t>();
-        par_for(output_shape.elements(), [&](auto i) {
-            auto data_idx          = output_shape.multi(i);
-            auto in_data_multi_idx = data_idx;
-            in_data_multi_idx[axis] *= 2;
-            auto input_val = input[in_data_multi_idx];
-            // mask first 4 bits, keep it little endian.
-            output[i] = uint8_t(0x0F) & input_val;
-            in_data_multi_idx[axis] += 1;
-            input_val = input[in_data_multi_idx];
-            output[i] = (input_val << 4u) | output[i]; // NOLINT(hicpp-signed-bitwise)
+
+        visit_all(result, input)([&](auto out, auto inp) {
+            par_for(output_shape.elements(), [&](auto i) {
+                using type = typename decltype(inp)::value_type;
```

Comment:
You should check the size and throw:

```cpp
if constexpr(sizeof(type) != 1)
    MIGRAPHX_THROW("Unsupported type");
```

Reply:
Isn't it enough to check it in `normalize_compute_shape`?

Reply:
Actually we don't need this, since with the modified logic we do a visit_all() to handle all types. Thanks.